akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -161,14 +160,24 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
                              connectionContext: ConnectionContext = defaultServerHttpContext,
                              settings:          ServerSettings    = ServerSettings(system),
                              log:               LoggingAdapter    = system.log): Source[Http.IncomingConnection, Future[ServerBinding]] = {
-    val fullLayer = fuseServerBidiFlow(settings, connectionContext, log)
+    val fullLayer: ServerLayerBidiFlow = fuseServerBidiFlow(settings, connectionContext, log)
 
     tcpBind(interface, choosePort(port, connectionContext, settings), settings)
       .map(incoming ⇒ {
-        val serverFlow = fullLayer.addAttributes(prepareAttributes(settings, incoming)) join incoming.flow
+        val preparedLayer: BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, ServerTerminator] = fullLayer.addAttributes(prepareAttributes(settings, incoming))
+        val serverFlow: Flow[HttpResponse, HttpRequest, ServerTerminator] = preparedLayer join incoming.flow
         IncomingConnection(incoming.localAddress, incoming.remoteAddress, serverFlow)
       })
-      .mapMaterializedValue(materializeTcpBind)
+      .mapMaterializedValue {
+        _.map(tcpBinding ⇒
+          ServerBinding(tcpBinding.localAddress)(
+            () ⇒ tcpBinding.unbind(),
+            deadline ⇒ {
+              ??? // FIXME uhm... do the same terminator dance here I guess
```

Comment:
FIXing this now, sadly a bit of setup dance has to be replicated from the other place